tools/iree-prof-output-stdout.cc

```diff
@@ -22,114 +22,285 @@
 namespace iree_prof {
 namespace {
 
+const char* ArchToString(tracy::CpuArchitecture arch) {
+  switch (arch) {
+    case tracy::CpuArchUnknown: return "Unknown";
+    case tracy::CpuArchX86: return "x86";
+    case tracy::CpuArchX64: return "x86_64";
+    case tracy::CpuArchArm32: return "arm";
+    case tracy::CpuArchArm64: return "aarch64";
+    default: return "Unknown";
+  }
+}
+
 bool HasSubstr(absl::string_view str, const std::vector<std::string>& substrs) {
   return std::find_if(
              substrs.begin(), substrs.end(),
              [str](const std::string& s) { return str.find(s) != str.npos; })
              != substrs.end();
 }
 
+std::string GetDurationStr(int64_t duration_ns,
+                           IreeProfOutputStdout::DurationUnit unit) {
+  switch (unit) {
+    case IreeProfOutputStdout::DurationUnit::kNanoseconds:
+      return absl::StrCat(duration_ns, "ns");
+    case IreeProfOutputStdout::DurationUnit::kMicroseconds:
+      return absl::StrCat(static_cast<double>(duration_ns) / 1000, "us");
+    case IreeProfOutputStdout::DurationUnit::kSeconds:
+      return absl::StrCat(static_cast<double>(duration_ns) / 1000000000, "s");
+    case IreeProfOutputStdout::DurationUnit::kMilliseconds:
+    default:
+      return absl::StrCat(static_cast<double>(duration_ns) / 1000000, "ms");
+  }
+}
+
+struct LongestDuration {
+  const char* zone_name;
+  int64_t duration;
+};
+
+template <typename T>
+absl::flat_hash_map<int, LongestDuration> GetLongestDurationPerThread(
+    const tracy::Worker& worker,
+    const tracy::unordered_flat_map<int16_t, T>& zones,
+    const std::vector<std::string>& thread_substrs) {
+  absl::flat_hash_map<int, LongestDuration> longest_durations;
+  TracyZoneFunctions<T> func;
+  for (const auto& z : zones) {
+    for (const auto& t : z.second.zones) {
+      auto tid = t.Thread();
+      auto duration = (t.Zone()->*func.end)() - (t.Zone()->*func.start)();
```

Comment:
What's the definition of the longest duration? Is it the duration between the start and end? If so, don't we need to get max(end) - min(start)?

Reply:
Done